File: cs/Morth/Parser.cs
namespace Morth;

public static class Parser
{
    public static Op ParseTokenAsOp(Token token)
    {
        if (token.Text == "+")
        {
            return Op.Plus(token.Location);
        }
        if (token.Text == "-")
        {
            return Op.Minus(token.Location);
        }
        if (token.Text == "dump")
        {
            return Op.Dump(token.Location);
        }
        if (token.Text == "=")
        {
            return Op.Equal(token.Location);
        }
        if (token.Text == "if")
        {
            return Op.If(token.Location);
        }
        if (token.Text == "else")
        {
            return Op.Else(token.Location);
        }
        if (token.Text == "end")
        {
            return Op.End(token.Location);
        }
        if (token.Text == "dup")
        {
            return Op.Dup(token.Location);
        }
        if (token.Text == ">")
        {
            return Op.Greater(token.Location);
        }
        if (token.Text == "while")
        {
            return Op.While(token.Location);
        }
        if (token.Text == "do")
        {
            return Op.Do(token.Location);
        }
        if (token.Text == "mem")
        {
            return Op.Mem(token.Location);
        }
        if (token.Text == ".")
        {
            return Op.Load(token.Location);
        }
        if (token.Text == ",")
        {
            return Op.Store(token.Location);
        }

        if (!ulong.TryParse(token.Text, out var value))
        {
            throw new ParseException(token, "Unknown token.");
        }

        return Op.Push(value, token.Location);
    }
}

File: fs/Morth.Language/Parser.fs
module Morth.Language.Parser

let parseWord (token : Token.t) =
  let code =
    (match token.word with
     | "+" -> Op.Plus
     | "-" -> Op.Minus
     | "=" -> Op.Eq
     | ">" -> Op.Gt
     | "dump" -> Op.Dump
     | "if" -> Op.If -1
     | "else" -> Op.Else -1
     | "while" -> Op.While
     | "do" -> Op.Do -1
     | "end" -> Op.End -1
     | "dup" -> Op.Dup
     | "mem" -> Op.Mem
     | "," -> Op.Load
     | "." -> Op.Store
     | "syscall1" -> Op.Syscall1
     | "syscall3" -> Op.Syscall3
     | n ->
       try
         Op.Push(int n)
       with err ->
         Printf.eprintfn "%s: %s" (token.loc.ToString()) err.Message
         exit 1) in

  new Op.t (token.loc, code)

let parse (filePath : string) =
  Lexer.lexFile filePath |> Seq.map parseWord |> Blocks.resolve
